src/Database/LSMTree/Internal/IndexOrdinaryAcc.hs

```diff
@@ -0,0 +1,100 @@
+{- HLINT ignore "Avoid restricted alias" -}
+
+{-|
+    Incremental construction functionality for the general-purpose fence pointer
+    index.
+-}
+module Database.LSMTree.Internal.IndexOrdinaryAcc
+(
+    IndexOrdinaryAcc,
+    new,
+    append,
+    end
+)
+where
+
+import           Control.Monad.ST.Strict (ST)
+import           Data.List (genericReplicate)
+import           Data.Primitive.ByteArray (byteArrayFromListN)
+import           Data.STRef.Strict (STRef, modifySTRef, newSTRef, readSTRef)
+import           Data.Vector (fromList)
+import qualified Data.Vector.Primitive as Primitive (Vector, length)
+import           Data.Word (Word16, Word8)
+import           Database.LSMTree.Internal.Chunk (Baler, Chunk, createBaler,
+                     feedBaler, readBalerRemnant)
+import           Database.LSMTree.Internal.IndexCompactAcc
+                     (Append (AppendMultiPage, AppendSinglePage))
+import           Database.LSMTree.Internal.IndexOrdinary
+                     (IndexOrdinary (IndexOrdinary))
+import           Database.LSMTree.Internal.Serialise
+                     (SerialisedKey (SerialisedKey'))
+import           Database.LSMTree.Internal.Vector (mkPrimVector)
+
+{-|
+    A general-purpose fence pointer index under incremental construction.
+
+    A value @IndexOrdinaryAcc lastKeysRevRef baler@ denotes a partially
+    constructed index with the following properties:
+
+      * The keys pointed to by @lastKeysRevRef@ are the keys that the index
+        assigns to pages, but in reverse order.
+      * The @baler@ object is used by the index for incremental output of the
+        serialised key list.
+-}
+data IndexOrdinaryAcc s = IndexOrdinaryAcc !(STRef s [SerialisedKey]) !(Baler s)
+
+-- | Creates a new, initially empty, index.
+new :: Int                       -- ^ Maximum chunk size in bytes
+    -> ST s (IndexOrdinaryAcc s) -- ^ Construction of the index
+new maxChunkSize = IndexOrdinaryAcc <$> newSTRef [] <*> createBaler maxChunkSize
+
+{-
+    Appends a single key to the key list of an index and outputs newly available
+    chunks of the serialised key list.
+-}
+appendKey :: SerialisedKey -> IndexOrdinaryAcc s -> ST s [Chunk]
+appendKey lastKey@(SerialisedKey' lastKeyBytes)
+          (IndexOrdinaryAcc lastKeysRevRef baler)
+    = do
+          modifySTRef lastKeysRevRef (lastKey :)
+          chunksIntoLastKeySize <- feedBaler lastKeySizeBytes baler
+          chunksIntoLastKey <- feedBaler lastKeyBytes baler
```

Comment:
One could write a `feedBaler2` function that gets two arguments, instead of calling `feedBaler` twice

Reply:
Hmm, having a function for the very special case of 2 blocks seems to go a bit the wrong way in my opinion. However, what about changing `feedBaler` such that it doesn’t accept a single block but a list of blocks?

I think that this change would be advisable anyhow. Currently, the baler implementation already respects #303 by having `feedBaler` output at most one chunk. However, the ordinary index can’t fully profit from this yet, because it calls `feedBaler` several times per appending. Having `feedBaler` accept an arbitrary number of blocks but still produce at most one chunk would enable `append` to also output at most one chunk, even in the case of a multi-page appending.

What are your thoughts on this?

Reply:
Hmm a list is probably nicest, yes. It would also mean that we can use that in the `AppendMultiPage` case

Reply:
The `feedBaler` function has been changed accordingly in `4d9463ff6f0c5f3cfafbcc4e82caa4a992229488`.

Reply:
With 916a5e84be04684ef87a3cf60b78233c2ef80361, `append` fully makes use of the extended `feedBaler` functionality, producing at most one chunk.